### ISSUE TITLE
AnnexRepo.add() ignores symlinks on Windows (GitRepo.add() works)

### ISSUE THREAD

**Issue by mih** [datalad/datalad#2955]
Demo:

This is a function from datalad-revolution that failed in https://ci.appveyor.com/project/mih/datalad-revolution/builds/19855520 in the analog of this test of `add`: https://github.com/datalad/datalad/blob/84f44bc882f01f4e3a4c28bff92a6cb7e6f1fc69/datalad/distribution/tests/test_add.py#L436
```
(Pdb) ds.repo.save()
*** datalad.support.exceptions.FileNotInRepositoryError: FileNotInRepositoryErro
r: command '['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'commit', '-m',
 'dryrun', '--dry-run', '--no-status', '--', 'symlink']' failed with exitcode 1
File(s) unknown to git
[Errno 1] ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'commit', '-m', '
dryrun', '--dry-run', '--no-status', '--', 'symlink']: File(s) unknown to git: ''
```
We are on windows:
```
(Pdb) path
'C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\datalad_temp_yhax7f_o'
```
The reason is that AnnexRepo.add() simply ignores the presence of the symlink, and does nothing (not error), and subsequently `Repo.commit()` fails, because it tries to commit something that had not been added yet.
```
(Pdb) ds.repo.add('symlink')
[]
```
In contrast, `GitRepo.add()` does the right thing (for this demo, the symlink was unstaged in a separate console session before this next line.
```
(Pdb) from datalad .support.gitrepo import GitRepo
(Pdb) GitRepo.add(ds.repo, files='symlink')
{'success': True, 'file': 'symlink'}
```

I am aware that this test is not even considered as part of the -core Windows tests, but -revolution is using this function still, and its failure seems to be the last issues before achieving full symlink capability on windows in -revolution -- hence reporting it here.

**Comment by mih** [comment]
This reason for this behavior is that `git annex add` itself ignores the symlink on Windows, but not on a real OS. Git itself shows uniform behavior on both platforms (demo below). This is relatively easy to work around within DataLad, but maybe @joeyh considers this a bug in git-annex?

Using V6:

```
C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git status
On branch adjusted/master(unlocked)
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        symlink

nothing added to commit but untracked files present (use "git add" to track)

C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git annex add symlink

C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git status
On branch adjusted/master(unlocked)
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        symlink

nothing added to commit but untracked files present (use "git add" to track)

C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git add symlink

C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git status
On branch adjusted/master(unlocked)
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        new file:   symlink


C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>

C:\Users\appveyor\AppData\Local\Temp\1\datalad_temp_lam28tjb>git annex version
git-annex version: 6.20180927-gc5b6c55af
build flags: Assistant Webapp Pairing S3(multipartupload)(storageclasses) WebDAV ConcurrentOutput TorrentParser Feeds Te
stsuite
dependency versions: aws-0.17.1 bloomfilter-2.0.1.0 cryptonite-0.23 DAV-1.3.1 feed-0.3.12.0 ghc-8.0.2 http-client-0.5.7.
1 persistent-sqlite-2.6.2 torrent-10000.1.1 uuid-1.3.13 yesod-1.4.5
key/value backends: SHA256E SHA256 SHA512E SHA512 SHA224E SHA224 SHA384E SHA384 SHA3_256E SHA3_256 SHA3_512E SHA3_512 SH
A3_224E SHA3_224 SHA3_384E SHA3_384 SKEIN256E SKEIN256 SKEIN512E SKEIN512 BLAKE2B256E BLAKE2B256 BLAKE2B512E BLAKE2B512
BLAKE2B160E BLAKE2B160 BLAKE2B224E BLAKE2B224 BLAKE2B384E BLAKE2B384 BLAKE2S256E BLAKE2S256 BLAKE2S160E BLAKE2S160 BLAKE
2S224E BLAKE2S224 BLAKE2SP256E BLAKE2SP256 BLAKE2SP224E BLAKE2SP224 SHA1E SHA1 MD5E MD5 WORM URL
remote types: git gcrypt p2p S3 bup directory rsync web bittorrent webdav adb tahoe glacier ddar hook external
operating system: mingw32 i386
supported repository versions: 3 5 6
upgrade supported from repository versions: 2 3 4 5
local repository version: 6
```

**Comment by joeyh** [comment]
Michael Hanke wrote:
> This reason for this behavior is that git annex add itself ignores the symlink
> on Windows, but not on a real OS. Git itself shows uniform behavior on both
> platforms (demo below). This is relatively easy to work around within DataLad,
> but maybe @joeyh considers this a bug in git-annex?

Is this an actual symlink or the dummy file git uses to stand in for it
on filesystems not supporting symlinks?

git-annex on Windows does not have access to any system calls for
operating on symlinks, AFAIK, so it would be hard for it to do anything
with them.

-- 
see shy jo

**Comment by mih** [comment]
This is an actual symlink. I have worked around this, by making a dedicated call to `git add` for symlinks on windows only.

**Comment by mih** [comment]
I am closing this now, because it has been fixed in master for a long time and 0.11 doesn't even get that far on Windows.
